### PR TITLE
Fixes #23633 - Use Object.assign to reduce loop length

### LINKED ISSUE #23633: [Feature]: Creating palette color objects should be $O(n)$ instead of $O(n^2)$
### Library

React Components / v9 (@fluentui/react-components)

### Describe the feature that you would like added

Status color palettes are created by looping through each shared color name ($n$) and building an object of theme names.

This loop is done [using a reduce + object spread call](https://github.com/microsoft/fluentui/blob/b00790b7a0ea1473d8c3cc49c7ca0088002957ed/packages/react-components/react-theme/src/alias/lightColorPalette.ts#L33), which creates a new copy of the object at each iteration. This makes the object creation $O(n^2)$ since it unnecessarily iterates over every key created so far. 

There are a fair number of color keys so this can create a performance issue when an app is starting up.

The easiest solution here is to use `Object.assign(acc, sharedColorTokens)` which will simply assign entries from `sharedColorTokens` into the existing `acc` object, mutating it. This makes the operation $O(n)$.

### Have you discussed this feature with our team

_No response_

### Additional context

https://twitter.com/jaffathecake/status/1396706523192115203

### Validations

- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

### PR DESCRIPTION
**Pull request by NotWoods**
## Current Behavior

A few `.reduce` loops do unnecessary work that increases the runtime when merging objects

## New Behavior

The loops have been updated to mutate `acc` rather than return a new object, which avoids the need to iterate over all existing keys of `acc`.

## Related Issue(s)

Fixes #23633

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/Avatar-mount_base.html">1342</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/Avatar-mount.html">1279</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/Button-mount_base.html">952</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/Button-mount.html">951</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProvider-mount_base.html">1589</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProvider-mount.html">1459</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProviderWithTheme-mount_base.html">557</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProviderWithTheme-mount.html">611</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">604</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">534</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">645</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">554</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/MakeStyles-mount_base.html">1785</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/MakeStyles-mount.html">1954</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/SpinButton-mount_base.html">2445</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/25433/perf-test-react-components/SpinButton-mount.html">2506</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                                                                |     Baseline (minified/GZIP) |                           PR |                                                                                                                                                                                                 Change |
| :--------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------: | ---------------------------: | -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> |  `62.94 kB`<br />`17.663 kB` | `62.968 kB`<br />`17.678 kB` | `28 B` <br />`15 B`  |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>               | `33.446 kB`<br />`11.033 kB` | `33.474 kB`<br />`11.042 kB` |  `28 B` <br />`9 B`  |
| <samp>react-theme</samp> <br /> <abbr title='bundle-size/TeamsAllThemes.fixture.js'>Teams: all themes</abbr>                                                     |   `29.65 kB`<br />`6.444 kB` |  `29.734 kB`<br />`6.464 kB` | `84 B` <br />`20 B`  |
| <samp>react-theme</samp> <br /> <abbr title='bundle-size/TeamsLightTheme.fixture.js'>Teams: Light theme</abbr>                                                   |  `17.486 kB`<br />`5.057 kB` |  `17.514 kB`<br />`5.076 kB` | `28 B` <br />`19 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>react-accordion</samp> <br /> <abbr title='bundle-size/Accordion.fixture.js'>Accordion (including children components)</abbr>                                           |  `79.243 kB`<br />`24.436 kB` |
| <samp>react-alert</samp> <br /> <abbr title='bundle-size/Alert.fixture.js'>Alert</abbr>                                                                                       |  `84.688 kB`<br />`21.233 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/Avatar.fixture.js'>Avatar</abbr>                                                                                    |  `48.874 kB`<br />`13.864 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/AvatarGroup.fixture.js'>AvatarGroup</abbr>                                                                          |   `14.996 kB`<br />`6.013 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/AvatarGroupItem.fixture.js'>AvatarGroupItem</abbr>                                                                  |  `63.452 kB`<br />`17.959 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/Badge.fixture.js'>Badge</abbr>                                                                                       |   `22.646 kB`<br />`7.228 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/CounterBadge.fixture.js'>CounterBadge</abbr>                                                                         |   `23.536 kB`<br />`7.512 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/PresenseBadge.fixture.js'>PresenceBadge</abbr>                                                                       |    `24.096 kB`<br />`7.09 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/Button.fixture.js'>Button</abbr>                                                                                    |   `36.849 kB`<br />`9.907 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/CompoundButton.fixture.js'>CompoundButton</abbr>                                                                    |  `43.874 kB`<br />`11.123 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/MenuButton.fixture.js'>MenuButton</abbr>                                                                            |  `39.543 kB`<br />`10.807 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/SplitButton.fixture.js'>SplitButton</abbr>                                                                          |  `46.989 kB`<br />`12.192 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/ToggleButton.fixture.js'>ToggleButton</abbr>                                                                        |  `52.716 kB`<br />`11.381 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.All.fixture.js'>Card - All</abbr>                                                                                |  `67.664 kB`<br />`19.635 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.fixture.js'>Card</abbr>                                                                                          |  `63.346 kB`<br />`18.539 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardFooter.fixture.js'>CardFooter</abbr>                                                                              |    `8.607 kB`<br />`3.622 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardHeader.fixture.js'>CardHeader</abbr>                                                                              |     `9.65 kB`<br />`3.962 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardPreview.fixture.js'>CardPreview</abbr>                                                                            |    `8.708 kB`<br />`3.678 kB` |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Combobox.fixture.js'>Combobox (including child components)</abbr>                                                 |   `75.85 kB`<br />`24.632 kB` |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Dropdown.fixture.js'>Dropdown (including child components)</abbr>                                                 |  `75.579 kB`<br />`24.594 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `189.603 kB`<br />`52.909 kB` |
| <samp>react-dialog</samp> <br /> <abbr title='bundle-size/Dialog.fixture.js'>Dialog (including children components)</abbr>                                                    |  `83.147 kB`<br />`24.799 kB` |
| <samp>react-divider</samp> <br /> <abbr title='bundle-size/Divider.fixture.js'>Divider</abbr>                                                                                 |   `16.505 kB`<br />`5.925 kB` |
| <samp>react-image</samp> <br /> <abbr title='bundle-size/Image.fixture.js'>Image</abbr>                                                                                       |   `10.826 kB`<br />`4.286 kB` |
| <samp>react-infobutton</samp> <br /> <abbr title='bundle-size/InfoButton.fixture.js'>InfoButton</abbr>                                                                        |    `7.742 kB`<br />`3.265 kB` |
| <samp>react-input</samp> <br /> <abbr title='bundle-size/Input.fixture.js'>Input</abbr>                                                                                       |   `24.428 kB`<br />`7.899 kB` |
| <samp>react-label</samp> <br /> <abbr title='bundle-size/Label.fixture.js'>Label</abbr>                                                                                       |    `9.384 kB`<br />`3.883 kB` |
| <samp>react-link</samp> <br /> <abbr title='bundle-size/Label.fixture.js'>Link</abbr>                                                                                         |   `11.852 kB`<br />`4.888 kB` |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.fixture.js'>Menu (including children components)</abbr>                                                          |  `116.78 kB`<br />`36.112 kB` |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.Selectable.fixture.js'>Menu (including selectable components)</abbr>                                             | `119.849 kB`<br />`36.635 kB` |
| <samp>react-overflow</samp> <br /> <abbr title='bundle-size/hooks.fixture.js'>hooks only</abbr>                                                                               |   `11.004 kB`<br />`4.188 kB` |
| <samp>react-persona</samp> <br /> <abbr title='bundle-size/Persona.fixture.js'>Persona</abbr>                                                                                 |   `53.992 kB`<br />`15.25 kB` |
| <samp>react-popover</samp> <br /> <abbr title='bundle-size/Popover.fixture.js'>Popover</abbr>                                                                                 |  `103.342 kB`<br />`31.82 kB` |
| <samp>react-portal</samp> <br /> <abbr title='bundle-size/Portal.fixture.js'>Portal</abbr>                                                                                    |   `10.628 kB`<br />`3.899 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |    `5.857 kB`<br />`1.978 kB` |
| <samp>react-positioning</samp> <br /> <abbr title='bundle-size/usePopper.fixture.js'>usePositioning</abbr>                                                                    |   `19.724 kB`<br />`7.415 kB` |
| <samp>react-progress</samp> <br /> <abbr title='bundle-size/Progress.fixture.js'>Progress</abbr>                                                                              |   `13.131 kB`<br />`5.039 kB` |
| <samp>react-provider</samp> <br /> <abbr title='bundle-size/FluentProvider.fixture.js'>FluentProvider</abbr>                                                                  |   `15.807 kB`<br />`5.909 kB` |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/Radio.fixture.js'>Radio</abbr>                                                                                       |  `36.436 kB`<br />`12.126 kB` |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/RadioGroup.fixture.js'>RadioGroup</abbr>                                                                             |   `14.294 kB`<br />`5.723 kB` |
| <samp>react-select</samp> <br /> <abbr title='bundle-size/Select.fixture.js'>Select</abbr>                                                                                    |   `23.255 kB`<br />`7.954 kB` |
| <samp>react-slider</samp> <br /> <abbr title='bundle-size/Slider.fixture.js'>Slider</abbr>                                                                                    |  `31.978 kB`<br />`10.149 kB` |
| <samp>react-spinbutton</samp> <br /> <abbr title='bundle-size/SpinButton.fixture.js'>SpinButton</abbr>                                                                        |  `45.826 kB`<br />`12.858 kB` |
| <samp>react-spinner</samp> <br /> <abbr title='bundle-size/Spinner.fixture.js'>Spinner</abbr>                                                                                 |   `20.023 kB`<br />`6.461 kB` |
| <samp>react-switch</samp> <br /> <abbr title='bundle-size/Switch.fixture.js'>Switch</abbr>                                                                                    |  `33.388 kB`<br />`10.578 kB` |
| <samp>react-text</samp> <br /> <abbr title='bundle-size/Text.Default.fixture.js'>Text - Default</abbr>                                                                        |   `11.828 kB`<br />`4.627 kB` |
| <samp>react-text</samp> <br /> <abbr title='bundle-size/Text.Wrappers.fixture.js'>Text - Wrappers</abbr>                                                                      |   `15.138 kB`<br />`5.064 kB` |
| <samp>react-textarea</samp> <br /> <abbr title='bundle-size/Textarea.fixture.js'>Textarea</abbr>                                                                              |   `25.687 kB`<br />`8.331 kB` |
| <samp>react-theme</samp> <br /> <abbr title='bundle-size/SingleThemeToken.fixture.js'>Single theme token import</abbr>                                                        |            `69 B`<br />`89 B` |
| <samp>react-tooltip</samp> <br /> <abbr title='bundle-size/Tooltip.fixture.js'>Tooltip</abbr>                                                                                 |  `41.718 kB`<br />`14.687 kB` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/8af3b4d39ad1649ceeb5a2797ad00aa3f0b1547c'>8af3b4d39ad1649ceeb5a2797ad00aa3f0b1547c</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/react-components/react-theme/src/alias/lightColorPalette.ts
Comment (spmonahan) [inline packages/react-components/react-theme/src/alias/lightColorPalette.ts:34]: I think these could all be simplified a tad bit more to

```ts
return Object.assign(acc, sharedColorTokens);
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#return_value

**Review by spmonahan — COMMENTED** [review]
This looks good to me but since I'm not strictly one of the codeowners I'll hold off on approving.

Could you add a before and after benchmark as well? Not strictly necessary since the issue does a good job of describing how this change improves perf but it's always nice to see the actual numbers :)

**Review by bsunderhus — APPROVED** [review]
Thx for the PR @NotWoods!

nit: as @spmonahan suggested, you could just return the assign invocation, to simplify a little more.

**Review by spmonahan — APPROVED** [review]

File: change/@fluentui-react-theme-6b37e9cc-8f6e-4c94-b79d-d42e1bddab74.json
Comment (ling1726) [inline change/@fluentui-react-theme-6b37e9cc-8f6e-4c94-b79d-d42e1bddab74.json:3]:
```suggestion
  "comment": "refactor: Use Object.assign to reduce loop runtime",
```

**Comment by fabricteam** [comment]
## [🕵 fluentuiv9 Open the Visual Regressions report to inspect the 2 screenshots]( https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-tool-fluent/index.html?organization=uifabric&project=fabricpublic&prId=25433&commitId=3e83a8f4ebeaa09a3094dc88a9ed0696d84d7783&env=prod)

✅ There was 0 screenshots added, 0 screenshots removed, 1749 screenshots unchanged, 0 screenshots with different dimensions and 2 screenshots with visible difference.


  <details><summary><b>unknown</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.size+active+badge.normal.chromium.png|10| Changed|
|Avatar Converged.size+active+ring-shadow.normal.chromium.png|5| Changed|

</details>


<div id="vrtCommentfluentuiv9"/>

File: packages/react-components/react-theme/src/alias/darkColorPalette.ts
Comment (behowell) [inline packages/react-components/react-theme/src/alias/darkColorPalette.ts:19]: (nit) These could be simplified even further to skip creating the `sharedColorTokens` object entirely, and just assign directly to `acc`. I don't think this would be a huge additional perf improvement, so it's optional to make this extra change.
```ts
acc[`colorPalette${color}Background1`] = statusSharedColors[sharedColor].shade40;
acc[`colorPalette${color}Background2`] = statusSharedColors[sharedColor].shade30;
acc[`colorPalette${color}Background3`] = statusSharedColors[sharedColor].primary;
acc[`colorPalette${color}Foreground1`] = statusSharedColors[sharedColor].tint30;
acc[`colorPalette${color}Foreground2`] = statusSharedColors[sharedColor].tint40;
acc[`colorPalette${color}Foreground3`] = statusSharedColors[sharedColor].tint20;
acc[`colorPalette${color}BorderActive`] = statusSharedColors[sharedColor].tint30;
acc[`colorPalette${color}Border1`] = statusSharedColors[sharedColor].primary;
acc[`colorPalette${color}Border2`] = statusSharedColors[sharedColor].tint20;
```

**Pull request merged**
